fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java

```diff
@@ -15,54 +15,78 @@
             }
         }
         codeMap.computeIfAbsent(code, k -> new HashSet<>()).add(parameter);
 
-        if (urlMap.containsKey(url)) {
-            SearchParameter previous = urlMap.get(url);
+        // for versioned search params, we store them in the canonicalMap twice:
+        // once with their version and once without it
+        if (canonicalMap.containsKey(url)) {
+            SearchParameter previous = canonicalMap.get(url);
             if (previous.getExpression() == null || previous.getExpression().equals(parameter.getExpression())) {
                 if (!code.equals(previous.getCode().getValue())) {
-                    log.info("SearchParameter with url '" + url + "' already exists with the same expression; "
+                    log.info("SearchParameter '" + url + "' already exists with the same expression; "
                             + "adding additional code '" + code + "'");
                 }
             } else {
-                log.warning("SearchParameter with url '" + url + "' already exists with a different expression;\n"
-                        + "replacing [id=" + previous.getId() + ", expression=" + previous.getExpression().getValue()
-                        + "] with [id=" + parameter.getId() + ", expression=" + parameter.getExpression().getValue() + "]");
+                log.info("SearchParameter '" + url + "' already exists with a different expression;\n"
+                        + "replacing [id=" + previous.getId() + ", version=" + previous.getVersion().getValue() + ", expression=" + previous.getExpression().getValue()
+                        + "] with [id=" + parameter.getId() + ", version=" + parameter.getVersion().getValue() + ", expression=" + parameter.getExpression().getValue() + "]");
+            }
+        }
+        canonicalMap.put(url, parameter);
+
+        if (version != null) {
+            String canonical = url + (version == null ? "" : "|" + version);
+            if (canonicalMap.containsKey(canonical)) {
+                SearchParameter previous = canonicalMap.get(canonical);
+                if (previous.getExpression() == null || previous.getExpression().equals(parameter.getExpression())) {
+                    if (!code.equals(previous.getCode().getValue())) {
+                        log.info("SearchParameter '" + canonical + "' already exists with the same expression; "
+                                + "adding additional code '" + code + "'");
+                    }
+                } else {
+                    log.warning("SearchParameter '" + canonical + "' already exists with a different expression;\n"
+                            + "replacing [id=" + previous.getId() + ", expression=" + previous.getExpression().getValue()
+                            + "] with [id=" + parameter.getId() + ", expression=" + parameter.getExpression().getValue() + "]");
+                }
```

Comment:
Rather than duplicate this segment, maybe this deserves a helper method to limit the additional complexity in updating / maintaining ?

Reply:
the two uses here are very similar but there is a subtle difference which is that the first stanza will print the version (if there is one) whereas this second block is only entered when canonical includes the version and so I removed that from the `[ ]` content.

Reply:
Right no worries - more of a consistency check point.   LGTM